File: go/interpreter/lfvm/opcode.go
// Copyright (c) 2024 Fantom Foundation
//
// Use of this software is governed by the Business Source License included
// in the LICENSE file and at fantom.foundation/bsl11.
//
// Change Date: 2028-4-16
//
// On the date above, in accordance with the Business Source License, use of
// this software will be governed by the GNU Lesser General Public License v3.

package lfvm

import (
	"fmt"

	"github.com/Fantom-foundation/Tosca/go/tosca/vm"
)

type OpCode uint16

// opCodeMask defines the relevant trailing bits of an OpCode. Any two OpCodes
// with the same value when masked with opCodeMask are considered equal.
//
// The motivation for this is that the long-form EVM has a number of OpCodes
// that are not part of the original EVM. For those, values beyond the range
// [0-255] of the EVM's single-byte OpCodes are used. To that end, the OpCode
// data type in the LFVM is increased to 16 bits. However, in several places
// maps from LFVM OpCodes to properties are required to provide efficient
// lookup tables for properties. To avoid the need to maintain tables of
// 2^16 entries, the number of relevant bits is reduced to 9. Any leading bits
// are ignored when comparing OpCodes.
const opCodeMask = 0x1ff

// numOpCodes is the maximum number of OpCodes that can be defined.
const numOpCodes = opCodeMask + 1

// The following constants define the original EVM OpCodes, in the lfvm OpCode space.
const (
	// Stack operations
	POP    = OpCode(vm.POP)
	PUSH0  = OpCode(vm.PUSH0)
	PUSH1  = OpCode(vm.PUSH1)
	PUSH2  = OpCode(vm.PUSH2)
	PUSH3  = OpCode(vm.PUSH3)
	PUSH4  = OpCode(vm.PUSH4)
	PUSH5  = OpCode(vm.PUSH5)
	PUSH6  = OpCode(vm.PUSH6)
	PUSH7  = OpCode(vm.PUSH7)
	PUSH8  = OpCode(vm.PUSH8)
	PUSH9  = OpCode(vm.PUSH9)
	PUSH10 = OpCode(vm.PUSH10)
	PUSH11 = OpCode(vm.PUSH11)
	PUSH12 = OpCode(vm.PUSH12)
	PUSH13 = OpCode(vm.PUSH13)
	PUSH14 = OpCode(vm.PUSH14)
	PUSH15 = OpCode(vm.PUSH15)
	PUSH16 = OpCode(vm.PUSH16)
	PUSH17 = OpCode(vm.PUSH17)
	PUSH18 = OpCode(vm.PUSH18)
	PUSH19 = OpCode(vm.PUSH19)
	PUSH20 = OpCode(vm.PUSH20)
	PUSH21 = OpCode(vm.PUSH21)
	PUSH22 = OpCode(vm.PUSH22)
	PUSH23 = OpCode(vm.PUSH23)
	PUSH24 = OpCode(vm.PUSH24)
	PUSH25 = OpCode(vm.PUSH25)
	PUSH26 = OpCode(vm.PUSH26)
	PUSH27 = OpCode(vm.PUSH27)
	PUSH28 = OpCode(vm.PUSH28)
	PUSH29 = OpCode(vm.PUSH29)
	PUSH30 = OpCode(vm.PUSH30)
	PUSH31 = OpCode(vm.PUSH31)
	PUSH32 = OpCode(vm.PUSH32)

	DUP1  = OpCode(vm.DUP1)
	DUP2  = OpCode(vm.DUP2)
	DUP3  = OpCode(vm.DUP3)
	DUP4  = OpCode(vm.DUP4)
	DUP5  = OpCode(vm.DUP5)
	DUP6  = OpCode(vm.DUP6)
	DUP7  = OpCode(vm.DUP7)
	DUP8  = OpCode(vm.DUP8)
	DUP9  = OpCode(vm.DUP9)
	DUP10 = OpCode(vm.DUP10)
	DUP11 = OpCode(vm.DUP11)
	DUP12 = OpCode(vm.DUP12)
	DUP13 = OpCode(vm.DUP13)
	DUP14 = OpCode(vm.DUP14)
	DUP15 = OpCode(vm.DUP15)
	DUP16 = OpCode(vm.DUP16)

	SWAP1  = OpCode(vm.SWAP1)
	SWAP2  = OpCode(vm.SWAP2)
	SWAP3  = OpCode(vm.SWAP3)
	SWAP4  = OpCode(vm.SWAP4)
	SWAP5  = OpCode(vm.SWAP5)
	SWAP6  = OpCode(vm.SWAP6)
	SWAP7  = OpCode(vm.SWAP7)
	SWAP8  = OpCode(vm.SWAP8)
	SWAP9  = OpCode(vm.SWAP9)
	SWAP10 = OpCode(vm.SWAP10)
	SWAP11 = OpCode(vm.SWAP11)
	SWAP12 = OpCode(vm.SWAP12)
	SWAP13 = OpCode(vm.SWAP13)
	SWAP14 = OpCode(vm.SWAP14)
	SWAP15 = OpCode(vm.SWAP15)
	SWAP16 = OpCode(vm.SWAP16)

	// Control flow
	JUMP     = OpCode(vm.JUMP)
	JUMPI    = OpCode(vm.JUMPI)
	JUMPDEST = OpCode(vm.JUMPDEST)
	RETURN   = OpCode(vm.RETURN)
	REVERT   = OpCode(vm.REVERT)
	PC       = OpCode(vm.PC)
	STOP     = OpCode(vm.STOP)

	// Arithmetic
	ADD        = OpCode(vm.ADD)
	SUB        = OpCode(vm.SUB)
	MUL        = OpCode(vm.MUL)
	DIV        = OpCode(vm.DIV)
	SDIV       = OpCode(vm.SDIV)
	MOD        = OpCode(vm.MOD)
	SMOD       = OpCode(vm.SMOD)
	ADDMOD     = OpCode(vm.ADDMOD)
	MULMOD     = OpCode(vm.MULMOD)
	EXP        = OpCode(vm.EXP)
	SIGNEXTEND = OpCode(vm.SIGNEXTEND)

	// Complex function
	SHA3 = OpCode(vm.SHA3)

	// Comparison operations
	LT     = OpCode(vm.LT)
	GT     = OpCode(vm.GT)
	SLT    = OpCode(vm.SLT)
	SGT    = OpCode(vm.SGT)
	EQ     = OpCode(vm.EQ)
	ISZERO = OpCode(vm.ISZERO)

	// Bit-pattern operations
	AND  = OpCode(vm.AND)
	OR   = OpCode(vm.OR)
	XOR  = OpCode(vm.XOR)
	NOT  = OpCode(vm.NOT)
	BYTE = OpCode(vm.BYTE)
	SHL  = OpCode(vm.SHL)
	SHR  = OpCode(vm.SHR)
	SAR  = OpCode(vm.SAR)

	// Memory
	MSTORE  = OpCode(vm.MSTORE)
	MSTORE8 = OpCode(vm.MSTORE8)
	MLOAD   = OpCode(vm.MLOAD)
	MSIZE   = OpCode(vm.MSIZE)
	MCOPY   = OpCode(vm.MCOPY)

	// Storage
	SLOAD  = OpCode(vm.SLOAD)
	SSTORE = OpCode(vm.SSTORE)
	TLOAD  = OpCode(vm.TLOAD)
	TSTORE = OpCode(vm.TSTORE)

	// LOG
	LOG0 = OpCode(vm.LOG0)
	LOG1 = OpCode(vm.LOG1)
	LOG2 = OpCode(vm.LOG2)
	LOG3 = OpCode(vm.LOG3)
	LOG4 = OpCode(vm.LOG4)

	// System level instructions.
	ADDRESS        = OpCode(vm.ADDRESS)
	BALANCE        = OpCode(vm.BALANCE)
	ORIGIN         = OpCode(vm.ORIGIN)
	CALLER         = OpCode(vm.CALLER)
	CALLVALUE      = OpCode(vm.CALLVALUE)
	CALLDATALOAD   = OpCode(vm.CALLDATALOAD)
	CALLDATASIZE   = OpCode(vm.CALLDATASIZE)
	CALLDATACOPY   = OpCode(vm.CALLDATACOPY)
	CODESIZE       = OpCode(vm.CODESIZE)
	CODECOPY       = OpCode(vm.CODECOPY)
	GASPRICE       = OpCode(vm.GASPRICE)
	EXTCODESIZE    = OpCode(vm.EXTCODESIZE)
	EXTCODECOPY    = OpCode(vm.EXTCODECOPY)
	RETURNDATASIZE = OpCode(vm.RETURNDATASIZE)
	RETURNDATACOPY = OpCode(vm.RETURNDATACOPY)
	EXTCODEHASH    = OpCode(vm.EXTCODEHASH)
	CREATE         = OpCode(vm.CREATE)
	CALL           = OpCode(vm.CALL)
	CALLCODE       = OpCode(vm.CALLCODE)
	DELEGATECALL   = OpCode(vm.DELEGATECALL)
	CREATE2        = OpCode(vm.CREATE2)
	STATICCALL     = OpCode(vm.STATICCALL)
	SELFDESTRUCT   = OpCode(vm.SELFDESTRUCT)

	// Blockchain instructions
	BLOCKHASH   = OpCode(vm.BLOCKHASH)
	COINBASE    = OpCode(vm.COINBASE)
	TIMESTAMP   = OpCode(vm.TIMESTAMP)
	NUMBER      = OpCode(vm.NUMBER)
	PREVRANDAO  = OpCode(vm.PREVRANDAO)
	GAS         = OpCode(vm.GAS)
	GASLIMIT    = OpCode(vm.GASLIMIT)
	CHAINID     = OpCode(vm.CHAINID)
	SELFBALANCE = OpCode(vm.SELFBALANCE)
	BASEFEE     = OpCode(vm.BASEFEE)
	BLOBHASH    = OpCode(vm.BLOBHASH)
	BLOBBASEFEE = OpCode(vm.BLOBBASEFEE)

	// Invalid instruction
	INVALID = OpCode(vm.INVALID)
)

// The following constants define the extended set of OpCodes for the long-form
// EVM.
// These opcodes are specific to the long-form EVM and are not part of the
// original EVM.
const (
	// long-form EVM special instructions

	// JUMP_TO is a special instruction that is used to jump to the end of the
	// current basic block.
	//
	// Since due to the usage of immediate arguments in instructions like PUSH2
	// the code size of basic blocks can shrink compared to the original EVM,
	// gaps can appear between the end of a basic block and the beginning of the
	// next one indicated by a JUMPDEST instruction. Since all JUMPDEST
	// instructions have to remain at the same position in the code as in the
	// original EVM code, since jump-destinations of JUMP and JUMPI  operations
	// are computed dynamically, these gaps have to be filled with NOOP
	// instructions. To avoid having to process long sequences of NOOPs,
	// JUMP_TO instructions are used to skip them in a single step.
	//
	// The following restrictions are imposed on JUMP_TO instructions:
	//  - they must target the immediate succeeding JUMPDEST instruction
	//  - all instructions between the JUMP_TO and the JUMPDEST must be NOOPs
	//
	// These restrictions are enforced during the EVM to LFVM code conversion.
	JUMP_TO OpCode = iota + 0x100

	// NOOP is a special instruction that does nothing. It is used as a filler
	// instruction to pad basic blocks to the correct size.
	NOOP

	// DATA is a special instruction that is used to extend the size of OpCodes
	// that require more than the available 2-byte immediate arguments.
	// For instance, [PUSH4, 1, 2, 3, 4] in the original EVM code gets converted
	// to [(PUSH4, 1<<8 | 2),(DATA, 3<<8 | 4)].
	// Since DATA is marked explicitly as such, jump-destination checks can be
	// conducted in O(1) by checking the OpCode of an instruction. In the
	// implicit data encoding of EVM byte code, this would require a linear
	// search (which could be cached to amortize costs).
	DATA

	// Super-instructions
	SWAP2_SWAP1_POP_JUMP
	SWAP1_POP_SWAP2_SWAP1
	POP_SWAP2_SWAP1_POP
	POP_POP
	PUSH1_SHL
	PUSH1_ADD
	PUSH1_DUP1
	PUSH2_JUMP
	PUSH2_JUMPI

	PUSH1_PUSH1
	SWAP1_POP
	POP_JUMP
	SWAP2_SWAP1
	SWAP2_POP
	DUP2_MSTORE
	DUP2_LT

	ISZERO_PUSH2_JUMPI
	PUSH1_PUSH4_DUP3
	AND_SWAP1_POP_SWAP2_SWAP1
	PUSH1_PUSH1_PUSH1_SHL_SUB

	// _highestOpCode is an alias for the OpCode with the highest defined
	// numeric value. It is only intended to be used in the unit tests
	// associated to this OpCode definition file to verify that the OpCode
	// bit mask limit has not been exceeded.
	_highestOpCode = PUSH1_PUSH1_PUSH1_SHL_SUB
)

var toString = map[OpCode]string{
	DATA:    "DATA",
	NOOP:    "NOOP",
	JUMP_TO: "JUMP_TO",

	SWAP2_SWAP1_POP_JUMP:  "SWAP2_SWAP1_POP_JUMP",
	SWAP1_POP_SWAP2_SWAP1: "SWAP1_POP_SWAP2_SWAP1",
	POP_SWAP2_SWAP1_POP:   "POP_SWAP2_SWAP1_POP",
	PUSH2_JUMP:            "PUSH2_JUMP",
	PUSH2_JUMPI:           "PUSH2_JUMPI",
	DUP2_MSTORE:           "DUP2_MSTORE",
	DUP2_LT:               "DUP2_LT",

	SWAP1_POP:   "SWAP1_POP",
	POP_JUMP:    "POP_JUMP",
	SWAP2_SWAP1: "SWAP2_SWAP1",
	SWAP2_POP:   "SWAP2_POP",
	PUSH1_PUSH1: "PUSH1_PUSH1",
	PUSH1_ADD:   "PUSH1_ADD",
	PUSH1_DUP1:  "PUSH1_DUP1",
	POP_POP:     "POP_POP",
	PUSH1_SHL:   "PUSH1_SHL",

	ISZERO_PUSH2_JUMPI:        "ISZERO_PUSH2_JUMPI",
	PUSH1_PUSH4_DUP3:          "PUSH1_PUSH4_DUP3",
	AND_SWAP1_POP_SWAP2_SWAP1: "AND_SWAP1_POP_SWAP2_SWAP1",
	PUSH1_PUSH1_PUSH1_SHL_SUB: "PUSH1_PUSH1_PUSH1_SHL_SUB",
}

// String returns the string representation of the OpCode.
func (o OpCode) String() string {
	if o <= 0xFF {
		return vm.OpCode(o).String()
	}

	if str, ok := toString[o]; ok {
		return str
	}
	return fmt.Sprintf("op(0x%04X)", int16(o))
}

// HasArgument returns true if the second 16-bit word of the instruction is
// argument data.
func (o OpCode) HasArgument() bool {
	if PUSH1 <= o && o <= PUSH32 {
		return true
	}
	switch o {
	case DATA:
		return true
	case JUMP_TO:
		return true
	}
	if o.isSuperInstruction() {
		for _, subOp := range o.decompose() {
			if subOp.HasArgument() {
				return true
			}
		}
	}
	return false
}

func (o OpCode) isBaseInstruction() bool {
	return o < 0x100
}

func (o OpCode) isSuperInstruction() bool {
	return o.decompose() != nil
}

func (o OpCode) decompose() []OpCode {
	switch o {
	case SWAP2_SWAP1_POP_JUMP:
		return []OpCode{SWAP2, SWAP1, POP, JUMP}
	case SWAP1_POP_SWAP2_SWAP1:
		return []OpCode{SWAP1, POP, SWAP2, SWAP1}
	case POP_SWAP2_SWAP1_POP:
		return []OpCode{POP, SWAP2, SWAP1, POP}
	case POP_POP:
		return []OpCode{POP, POP}
	case PUSH1_SHL:
		return []OpCode{PUSH1, SHL}
	case PUSH1_ADD:
		return []OpCode{PUSH1, ADD}
	case PUSH1_DUP1:
		return []OpCode{PUSH1, DUP1}
	case PUSH2_JUMP:
		return []OpCode{PUSH2, JUMP}
	case PUSH2_JUMPI:
		return []OpCode{PUSH2, JUMPI}
	case PUSH1_PUSH1:
		return []OpCode{PUSH1, PUSH1}
	case SWAP1_POP:
		return []OpCode{SWAP1, POP}
	case POP_JUMP:
		return []OpCode{POP, JUMP}
	case SWAP2_SWAP1:
		return []OpCode{SWAP2, SWAP1}
	case SWAP2_POP:
		return []OpCode{SWAP2, POP}
	case DUP2_MSTORE:
		return []OpCode{DUP2, MSTORE}
	case DUP2_LT:
		return []OpCode{DUP2, LT}
	case ISZERO_PUSH2_JUMPI:
		return []OpCode{ISZERO, PUSH2, JUMPI}
	case PUSH1_PUSH4_DUP3:
		return []OpCode{PUSH1, PUSH4, DUP3}
	case AND_SWAP1_POP_SWAP2_SWAP1:
		return []OpCode{AND, SWAP1, POP, SWAP2, SWAP1}
	case PUSH1_PUSH1_PUSH1_SHL_SUB:
		return []OpCode{PUSH1, PUSH1, PUSH1, SHL, SUB}
	}
	return nil
}

// opCodePropertyMap is a generic property map for precomputed values.
// Its purpose is to provide a precomputed lookup table for OpCode properties
// that can be generated from a function that takes an OpCode as input.
// Using this type hides internal details of the opcode implementation.
type opCodePropertyMap[T any] struct {
	lookup [numOpCodes]T
}

// newOpCodePropertyMap creates a new OpCode property map.
// The property function shall be resilient to undefined OpCode values, and not
// panic. The zero values or a sentinel value shall be used in such cases.
func newOpCodePropertyMap[T any](property func(op OpCode) T) opCodePropertyMap[T] {
	lookup := [numOpCodes]T{}
	for i := 0; i < numOpCodes; i++ {
		lookup[i] = property(OpCode(i))
	}
	return opCodePropertyMap[T]{lookup}
}

func (p *opCodePropertyMap[T]) get(op OpCode) T {
	// Index may be out of bounds. Nevertheless, bounds check carry a performance
	// penalty. If the property map is initialized correctly, the index will be
	// within bounds.
	return p.lookup[op&opCodeMask]
}

File: rust/src/types/opcode.rs
const STOP: u8 = 0x00;
const ADD: u8 = 0x01;
const MUL: u8 = 0x02;
const SUB: u8 = 0x03;
const DIV: u8 = 0x04;
const SDIV: u8 = 0x05;
const MOD: u8 = 0x06;
const SMOD: u8 = 0x07;
const ADDMOD: u8 = 0x08;
const MULMOD: u8 = 0x09;
const EXP: u8 = 0x0A;
const SIGNEXTEND: u8 = 0x0B;
const LT: u8 = 0x10;
const GT: u8 = 0x11;
const SLT: u8 = 0x12;
const SGT: u8 = 0x13;
const EQ: u8 = 0x14;
const ISZERO: u8 = 0x15;
const AND: u8 = 0x16;
const OR: u8 = 0x17;
const XOR: u8 = 0x18;
const NOT: u8 = 0x19;
const BYTE: u8 = 0x1A;
const SHL: u8 = 0x1B;
const SHR: u8 = 0x1C;
const SAR: u8 = 0x1D;
const SHA3: u8 = 0x20;
const ADDRESS: u8 = 0x30;
const BALANCE: u8 = 0x31;
const ORIGIN: u8 = 0x32;
const CALLER: u8 = 0x33;
const CALLVALUE: u8 = 0x34;
const CALLDATALOAD: u8 = 0x35;
const CALLDATASIZE: u8 = 0x36;
const CALLDATACOPY: u8 = 0x37;
const CODESIZE: u8 = 0x38;
const CODECOPY: u8 = 0x39;
const GASPRICE: u8 = 0x3A;
const EXTCODESIZE: u8 = 0x3B;
const EXTCODECOPY: u8 = 0x3C;
const RETURNDATASIZE: u8 = 0x3D;
const RETURNDATACOPY: u8 = 0x3E;
const EXTCODEHASH: u8 = 0x3F;
const BLOCKHASH: u8 = 0x40;
const COINBASE: u8 = 0x41;
const TIMESTAMP: u8 = 0x42;
const NUMBER: u8 = 0x43;
const PREVRANDAO: u8 = 0x44;
const GASLIMIT: u8 = 0x45;
const CHAINID: u8 = 0x46;
const SELFBALANCE: u8 = 0x47;
const BASEFEE: u8 = 0x48;
const BLOBHASH: u8 = 0x49;
const BLOBBASEFEE: u8 = 0x4A;
const POP: u8 = 0x50;
const MLOAD: u8 = 0x51;
const MSTORE: u8 = 0x52;
const MSTORE8: u8 = 0x53;
const SLOAD: u8 = 0x54;
const SSTORE: u8 = 0x55;
const JUMP: u8 = 0x56;
const JUMPI: u8 = 0x57;
const PC: u8 = 0x58;
const MSIZE: u8 = 0x59;
const GAS: u8 = 0x5A;
const JUMPDEST: u8 = 0x5B;
const TLOAD: u8 = 0x5C;
const TSTORE: u8 = 0x5D;
const MCOPY: u8 = 0x5E;
const PUSH0: u8 = 0x5F;
const PUSH1: u8 = 0x60;
const PUSH2: u8 = 0x61;
const PUSH3: u8 = 0x62;
const PUSH4: u8 = 0x63;
const PUSH5: u8 = 0x64;
const PUSH6: u8 = 0x65;
const PUSH7: u8 = 0x66;
const PUSH8: u8 = 0x67;
const PUSH9: u8 = 0x68;
const PUSH10: u8 = 0x69;
const PUSH11: u8 = 0x6A;
const PUSH12: u8 = 0x6B;
const PUSH13: u8 = 0x6C;
const PUSH14: u8 = 0x6D;
const PUSH15: u8 = 0x6E;
const PUSH16: u8 = 0x6F;
const PUSH17: u8 = 0x70;
const PUSH18: u8 = 0x71;
const PUSH19: u8 = 0x72;
const PUSH20: u8 = 0x73;
const PUSH21: u8 = 0x74;
const PUSH22: u8 = 0x75;
const PUSH23: u8 = 0x76;
const PUSH24: u8 = 0x77;
const PUSH25: u8 = 0x78;
const PUSH26: u8 = 0x79;
const PUSH27: u8 = 0x7A;
const PUSH28: u8 = 0x7B;
const PUSH29: u8 = 0x7C;
const PUSH30: u8 = 0x7D;
const PUSH31: u8 = 0x7E;
const PUSH32: u8 = 0x7F;
const DUP1: u8 = 0x80;
const DUP2: u8 = 0x81;
const DUP3: u8 = 0x82;
const DUP4: u8 = 0x83;
const DUP5: u8 = 0x84;
const DUP6: u8 = 0x85;
const DUP7: u8 = 0x86;
const DUP8: u8 = 0x87;
const DUP9: u8 = 0x88;
const DUP10: u8 = 0x89;
const DUP11: u8 = 0x8A;
const DUP12: u8 = 0x8B;
const DUP13: u8 = 0x8C;
const DUP14: u8 = 0x8D;
const DUP15: u8 = 0x8E;
const DUP16: u8 = 0x8F;
const SWAP1: u8 = 0x90;
const SWAP2: u8 = 0x91;
const SWAP3: u8 = 0x92;
const SWAP4: u8 = 0x93;
const SWAP5: u8 = 0x94;
const SWAP6: u8 = 0x95;
const SWAP7: u8 = 0x96;
const SWAP8: u8 = 0x97;
const SWAP9: u8 = 0x98;
const SWAP10: u8 = 0x99;
const SWAP11: u8 = 0x9A;
const SWAP12: u8 = 0x9B;
const SWAP13: u8 = 0x9C;
const SWAP14: u8 = 0x9D;
const SWAP15: u8 = 0x9E;
const SWAP16: u8 = 0x9F;
const LOG0: u8 = 0xA0;
const LOG1: u8 = 0xA1;
const LOG2: u8 = 0xA2;
const LOG3: u8 = 0xA3;
const LOG4: u8 = 0xA4;
const CREATE: u8 = 0xF0;
const CALL: u8 = 0xF1;
const CALLCODE: u8 = 0xF2;
const RETURN: u8 = 0xF3;
const DELEGATECALL: u8 = 0xF4;
const CREATE2: u8 = 0xF5;
const STATICCALL: u8 = 0xFA;
const REVERT: u8 = 0xFD;
const INVALID: u8 = 0xFE;
const SELFDESTRUCT: u8 = 0xFF;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(u8)]
pub enum Opcode {
    Stop = STOP,
    Add = ADD,
    Mul = MUL,
    Sub = SUB,
    Div = DIV,
    SDiv = SDIV,
    Mod = MOD,
    SMod = SMOD,
    AddMod = ADDMOD,
    MulMod = MULMOD,
    Exp = EXP,
    SignExtend = SIGNEXTEND,
    Lt = LT,
    Gt = GT,
    SLt = SLT,
    SGt = SGT,
    Eq = EQ,
    IsZero = ISZERO,
    And = AND,
    Or = OR,
    Xor = XOR,
    Not = NOT,
    Byte = BYTE,
    Shl = SHL,
    Shr = SHR,
    Sar = SAR,
    Sha3 = SHA3,
    #[cfg(feature = "needs-fn-ptr-conversion")]
    NoOp = SHA3 + 1,
    #[cfg(feature = "needs-fn-ptr-conversion")]
    SkipNoOps = SHA3 + 2,
    Address = ADDRESS,
    Balance = BALANCE,
    Origin = ORIGIN,
    Caller = CALLER,
    CallValue = CALLVALUE,
    CallDataLoad = CALLDATALOAD,
    CallDataSize = CALLDATASIZE,
    CallDataCopy = CALLDATACOPY,
    CodeSize = CODESIZE,
    CodeCopy = CODECOPY,
    GasPrice = GASPRICE,
    ExtCodeSize = EXTCODESIZE,
    ExtCodeCopy = EXTCODECOPY,
    ReturnDataSize = RETURNDATASIZE,
    ReturnDataCopy = RETURNDATACOPY,
    ExtCodeHash = EXTCODEHASH,
    BlockHash = BLOCKHASH,
    Coinbase = COINBASE,
    Timestamp = TIMESTAMP,
    Number = NUMBER,
    PrevRandao = PREVRANDAO,
    GasLimit = GASLIMIT,
    ChainId = CHAINID,
    SelfBalance = SELFBALANCE,
    BaseFee = BASEFEE,
    BlobHash = BLOBHASH,
    BlobBaseFee = BLOBBASEFEE,
    Pop = POP,
    MLoad = MLOAD,
    MStore = MSTORE,
    MStore8 = MSTORE8,
    SLoad = SLOAD,
    SStore = SSTORE,
    Jump = JUMP,
    JumpI = JUMPI,
    Pc = PC,
    MSize = MSIZE,
    Gas = GAS,
    JumpDest = JUMPDEST,
    TLoad = TLOAD,
    TStore = TSTORE,
    MCopy = MCOPY,
    Push0 = PUSH0,
    Push1 = PUSH1,
    Push2 = PUSH2,
    Push3 = PUSH3,
    Push4 = PUSH4,
    Push5 = PUSH5,
    Push6 = PUSH6,
    Push7 = PUSH7,
    Push8 = PUSH8,
    Push9 = PUSH9,
    Push10 = PUSH10,
    Push11 = PUSH11,
    Push12 = PUSH12,
    Push13 = PUSH13,
    Push14 = PUSH14,
    Push15 = PUSH15,
    Push16 = PUSH16,
    Push17 = PUSH17,
    Push18 = PUSH18,
    Push19 = PUSH19,
    Push20 = PUSH20,
    Push21 = PUSH21,
    Push22 = PUSH22,
    Push23 = PUSH23,
    Push24 = PUSH24,
    Push25 = PUSH25,
    Push26 = PUSH26,
    Push27 = PUSH27,
    Push28 = PUSH28,
    Push29 = PUSH29,
    Push30 = PUSH30,
    Push31 = PUSH31,
    Push32 = PUSH32,
    Dup1 = DUP1,
    Dup2 = DUP2,
    Dup3 = DUP3,
    Dup4 = DUP4,
    Dup5 = DUP5,
    Dup6 = DUP6,
    Dup7 = DUP7,
    Dup8 = DUP8,
    Dup9 = DUP9,
    Dup10 = DUP10,
    Dup11 = DUP11,
    Dup12 = DUP12,
    Dup13 = DUP13,
    Dup14 = DUP14,
    Dup15 = DUP15,
    Dup16 = DUP16,
    Swap1 = SWAP1,
    Swap2 = SWAP2,
    Swap3 = SWAP3,
    Swap4 = SWAP4,
    Swap5 = SWAP5,
    Swap6 = SWAP6,
    Swap7 = SWAP7,
    Swap8 = SWAP8,
    Swap9 = SWAP9,
    Swap10 = SWAP10,
    Swap11 = SWAP11,
    Swap12 = SWAP12,
    Swap13 = SWAP13,
    Swap14 = SWAP14,
    Swap15 = SWAP15,
    Swap16 = SWAP16,
    Log0 = LOG0,
    Log1 = LOG1,
    Log2 = LOG2,
    Log3 = LOG3,
    Log4 = LOG4,
    Create = CREATE,
    Call = CALL,
    CallCode = CALLCODE,
    Return = RETURN,
    DelegateCall = DELEGATECALL,
    Create2 = CREATE2,
    StaticCall = STATICCALL,
    Revert = REVERT,
    Invalid = INVALID,
    SelfDestruct = SELFDESTRUCT,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CodeByteType {
    JumpDest,
    #[cfg(feature = "needs-fn-ptr-conversion")]
    Push,
    Opcode,
    DataOrInvalid,
}

pub fn code_byte_type(code_byte: u8) -> (CodeByteType, usize) {
    match code_byte {
        STOP | ADD | MUL | SUB | DIV | SDIV | MOD | SMOD | ADDMOD | MULMOD | EXP | SIGNEXTEND
        | LT | GT | SLT | SGT | EQ | ISZERO | AND | OR | XOR | NOT | BYTE | SHL | SHR | SAR
        | SHA3 | ADDRESS | BALANCE | ORIGIN | CALLER | CALLVALUE | CALLDATALOAD | CALLDATASIZE
        | CALLDATACOPY | CODESIZE | CODECOPY | GASPRICE | EXTCODESIZE | EXTCODECOPY
        | RETURNDATASIZE | RETURNDATACOPY | EXTCODEHASH | BLOCKHASH | COINBASE | TIMESTAMP
        | NUMBER | PREVRANDAO | GASLIMIT | CHAINID | SELFBALANCE | BASEFEE | BLOBHASH
        | BLOBBASEFEE | POP | MLOAD | MSTORE | MSTORE8 | SLOAD | SSTORE | JUMP | JUMPI | PC
        | MSIZE | GAS | TLOAD | TSTORE | MCOPY | PUSH0 | DUP1 | DUP2 | DUP3 | DUP4 | DUP5
        | DUP6 | DUP7 | DUP8 | DUP9 | DUP10 | DUP11 | DUP12 | DUP13 | DUP14 | DUP15 | DUP16
        | SWAP1 | SWAP2 | SWAP3 | SWAP4 | SWAP5 | SWAP6 | SWAP7 | SWAP8 | SWAP9 | SWAP10
        | SWAP11 | SWAP12 | SWAP13 | SWAP14 | SWAP15 | SWAP16 | LOG0 | LOG1 | LOG2 | LOG3
        | LOG4 | CREATE | CALL | CALLCODE | RETURN | DELEGATECALL | CREATE2 | STATICCALL
        | REVERT | INVALID | SELFDESTRUCT => (CodeByteType::Opcode, 0),
        PUSH1..=PUSH32 => (
            #[cfg(not(feature = "needs-fn-ptr-conversion"))]
            CodeByteType::Opcode,
            #[cfg(feature = "needs-fn-ptr-conversion")]
            CodeByteType::Push,
            (code_byte - Opcode::Push1 as u8 + 1) as usize,
        ),
        JUMPDEST => (CodeByteType::JumpDest, 0),
        _ => (CodeByteType::DataOrInvalid, 0),
    }
}
